### ISSUE TITLE
Sparse exogeneous model matrix support

### ISSUE THREAD

**Issue by GordStephen** [FixedEffects/FixedEffectModels.jl#9]
Now that DataFrames.jl supports generating sparse model matrices, it would be great to put them to use in this package, for cases when you have very high-dimensional fixed effects but also moderately high-dimensional exogenous categorical variables of interest (where the resulting dense covariance matrix would be large but still fit in memory) when working with a large number of observations.

After starting working on this, most of the issues I've run into so far have just been a matter of generalizing method signatures, but my lack of experience working with Cholesky factorizations is tripping me up when it comes to reproducing the right pivots and permutations in some of the more numerical code. Specifically, `basecol` is [implemented](https://github.com/matthieugomez/FixedEffectModels.jl/blob/472565fbff82e7eaaed06aefbde74828be34d06f/src/utils/basecol.jl#L67) as

``` julia
function basecol(X::Matrix{Float64}...)
    chol = cholfact!(crossprod(X...), :U, Val{true})
    ipermute!(diag(chol.factors) .> 0, chol.piv)
end
```

so `chol` is the upper part of the pivoted factorization I think? The sparse equivalent (I believe) is just `chol = cholfact(crossprod(X...))`, since the sparse factorization is pivoted by default. From there though, there are options to extract the factorization with or without the effects of the permutation matrix (`chol[:U]` vs `chol[:UP]`) - is this choice related to the inverse permutation on the second line of the current `basecol` implementation? What would a version of this function with support for sparse `X` look like (given `crossprod` were sparse-compatible)?

My apologies for the fact that I don't really know what I'm talking about! I hope this is at least somewhat coherent, and a reasonable thing to be asking about.

**Comment by GordStephen** [comment]
What I ended up using was 

``` julia
function basecol(X::SparseMatrixCSC{Float64}...)
    spchol = cholfact(crossprod(X...))
    ipermute!(diag(spchol) .> 0, spchol[:p])
end
```

The only other necessary change of note was to define

``` julia
Base.inv(x::Base.SparseArrays.CHOLMOD.Factor) = x |> sparse |> full |> inv
```

since the sparse `CHOLMOD.Factor` type doesn't have an inverse method defined.

I now have a hacky version of this package working that uses a sparse model matrix, without support for  instrument variables ~~or QR or Cholesky residual methods.~~

~~Unfortunately, as one would expect, the in-place LSMR `residualize!` methods are quite slow when working with sparse matrices... Since `partial_out` seems to work fine on large datasets, I suspect there's a way to avoid needing to work with sparse matrices here?~~

Update: Since `residualize!` ultimately runs column-by-column, what I've ended up doing is copying each model matrix column to a dense array before residualizing (this also reenables support for the QR and Cholesky methods). The dense results are then converted back to a sparse vector and stored until all columns have been residualized. At that point, they're `hcat`ted back together and the fields of the original sparse model matrix are manually mutated to match the newly created sparse matrix, so as to simulate the in-place updating that occurs in the dense case. Not the most elegant solution, but it's at least several times faster than the sparse-all-the-way-down approach. I'm certainly open to better ideas though!
